src/main/java/com/google/sps/servlets/ChatServlet.java

```diff
@@ -0,0 +1,118 @@
+import com.google.template.soy.SoyFileSet;
+import com.google.template.soy.tofu.SoyTofu;
+import java.io.File;
+import java.io.IOException;
+import java.util.Map;
+import java.util.Scanner;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+
+@WebServlet("/chat")
+public class ChatServlet extends HttpServlet{
+    private static final String MESSAGE_TEXT_PROPERTY = "message-text";
+    private static final String TIMESTAMP_PROPERTY = "timestamp";
+    private static final String MESSAGE_KIND = "Message";
+
+  /**
+   * When called, doGet will query all previously posted messages to the chat,
+   * and render the messages into a HTML template.
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Calls query on all entities of type Message
+    Query messageQuery = new Query(MESSAGE_KIND).addSort(TIMESTAMP_PROPERTY, SortDirection.ASCENDING);
+    PreparedQuery preparedMessageQuery = datastore.prepare(messageQuery);
+
+    // Creates list of the messages text
+    ImmutableList<String> messagesList = Streams.stream(preparedMessageQuery.asIterable()).map(message -> 
+        (String) message.getProperty(MESSAGE_TEXT_PROPERTY)).collect(toImmutableList());
+
+    // Data will be passed in as a list of messages in a map (needed for template)
+    ImmutableMap<String, ImmutableList<String>> data = ImmutableMap.of("messages", messagesList);
+
+    // File path starts in target/portfolio-1
+    SoyFileSet sfs = SoyFileSet
+        .builder()
+        .add(new File("../../src/main/java/templates/chat.soy"))
+        .build();
+    SoyTofu tofu = sfs.compileToTofu();
+
+    final String out = tofu.newRenderer("templates.chat.chatPage").setData(data).render();
+    response.getWriter().println(out);
+  }
+
+/**
+ * Is called by a button on the chat form. doPost takes the message and its
+ * attributes (such as timestamp and user) and stores them into Datastore.
+ */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    final String messageText = (String) request.getParameter(MESSAGE_TEXT_PROPERTY);
+    final long timestamp = System.currentTimeMillis();
+    
+    // Reads API Key and Referer from file 
+    File apiKeyFile = new File("../../keys.txt");
+    Scanner scanner = new Scanner(apiKeyFile);
+    final String API_KEY = scanner.nextLine();
+    final String REFERER = scanner.nextLine();
```

Comment:
What is the Referer?

Reply:
there were restrictions on the API key when I tried to use it, so I had to explicitly state the website that was requesting (i.e. cloudshell)